### PR TITLE
AS-592: return 404 from workspace-specific APIs if the workspace doesn't exist [risk: low]

### PR DESCRIPTION
**Pull request by davidangb**
This would eliminate the need for broadinstitute/rawls#1347.

In this PR, we change workspace- and data reference-related APIs, which all operate in the context of a single workspace, to always return a 404 if the workspace doesn't exist. It still returns a 403 if the workspace exists but you don't have permission; and returns an appropriate code (generally 200/204) if you do have permission.

In the case of 403,  it changes the error message from "User %s is not authorized to %s workspace %s or it does not exist" to "User %s is not authorized to %s workspace %s". 

For those who haven't followed all the conversation, we realize the 403/404 differentiation leaks existence of a workspace to those who don't have permission to see it. Because Workspace Manager is a low-level service, this is ok; we obscure  error messages at higher layers of the stack.

Do I need to duplicate the new tests using testrunner syntaxes?

### CODE REVIEW

**Comment by jaycarlton** [comment]
Setting up your tests in TestRunner would be appreciated. It could be a separate PR though.

File: src/main/java/bio/terra/workspace/db/WorkspaceDao.java [hunk 1]
```diff
@@ -124,6 +124,17 @@ public WorkspaceStage getWorkspaceStage(UUID workspaceId) {
     return WorkspaceStage.valueOf(jdbcTemplate.queryForObject(sql, params, String.class));
   }
 
+  // TODO: do all three variants of assertMcWorkspace belong in WorkspaceDAO?
+  private void assertMcWorkspace(WorkspaceStage stage, UUID workspaceId, String action) {
```
Comment (jaycarlton) [inline src/main/java/bio/terra/workspace/db/WorkspaceDao.java:128]: nit: It would be nice to make the actions an enum with a string description constant, but that can be another PR.

File: src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java [hunk 2]
```diff
@@ -53,7 +53,8 @@ public DataReferenceService(
   public DataReference getDataReference(
       UUID workspaceId, UUID referenceId, AuthenticatedUserRequest userReq) {
 
-    samService.workspaceAuthz(userReq, workspaceId, SamConstants.SAM_WORKSPACE_READ_ACTION);
+    workspaceService.authorizedGetWorkspace(
```
Comment (jaycarlton) [inline src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java:56]: For assertions that throw, I like to prefix the function with `validate`, though I don't know if there's a prevailing convention here.

More importantly, this method checks more than workspace read, so maybe `validateWorkspaceOperation()` would be better.

File: src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java [hunk 3]
```diff
@@ -78,7 +80,7 @@ public DataReference getDataReferenceByName(
   public DataReference createDataReference(
```
Comment (jaycarlton) [inline src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java:80]: For public methods, maybe add to the javadoc that we validate workspace visibility and modifiability first, then validate the data reference operation, etc.

File: src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java [hunk 4]
```diff
@@ -68,18 +68,42 @@ public UUID createWorkspace(WorkspaceRequest workspaceRequest, AuthenticatedUser
...
+   * @param userReq the user's authenticated request
+   * @param workspaceId id of the workspace in question
+   * @param action the action to authorize against the workspace
+   * @return the workspace, if it exists and the user is permitted to perform the specified action.
+   */
+  @Traced
+  public Workspace authorizedGetWorkspace(
+      AuthenticatedUserRequest userReq, UUID workspaceId, String action) {
+    Workspace workspace = workspaceDao.getWorkspace(workspaceId);
+    samService.workspaceAuthzOnly(userReq, workspaceId, action);
```
Comment (jaycarlton) [inline src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java:93]: why is the auth check after the `getWorkspace()` call here?

File: src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java [hunk 5]
```diff
@@ -97,20 +121,20 @@ public void deleteWorkspace(UUID id, AuthenticatedUserRequest userReq) {
...
-    workspaceDao.assertMcWorkspace(workspaceId, "createGoogleContext");
+    Workspace workspace =
+        authorizedGetWorkspace(userReq, workspaceId, SamConstants.SAM_WORKSPACE_WRITE_ACTION);
+    workspaceDao.assertMcWorkspace(workspace, "createGoogleContext");
     if (!workspaceDao.getCloudContext(workspaceId).googleProjectId().isEmpty()) {
       throw new DuplicateGoogleContextException(workspaceId);
     }
-    Optional<SpendProfileId> spendProfileId =
-        workspaceDao.getWorkspace(workspaceId).spendProfileId();
+    Optional<SpendProfileId> spendProfileId = workspace.spendProfileId();
```
Comment (jaycarlton) [inline src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java:137]: nit: `SpendProfileId spendProfileId = workspace.spendProfileId().orElseThrow(() -> new MissingSpend...())`

File: src/main/resources/api/service_openapi.yaml [hunk 6]
```diff
@@ -114,6 +118,8 @@ paths:
           $ref: '#/components/responses/BadRequest'
         '403':
           $ref: '#/components/responses/PermissionDenied'
+        '404':
```
Comment (jaycarlton) [inline src/main/resources/api/service_openapi.yaml:121]: aside: I wish OpenAPI gave us a way to avoid re-declaring the map between error and type everywhere.

**Review by jaycarlton — APPROVED** [review]

**Review by zloery — APPROVED** [review]
LGTM
You shouldn't need to do anything for TestRunner - all the tests you touch here are unit and connected tests, and TestRunner is only for client library tests.

File: src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java (hunk 2 above)
Comment (davidangb) [reply src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java:56]: I hate naming! I want to also indicate in the name that this method retrieves and returns a `Workspace` object, after validating permissions. I'll try to come up with something better.

File: src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java (hunk 4 above)
Comment (davidangb) [reply src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java:93]: this is the core of the entire PR. As discussed in Slack, consensus is that we always want to return 404 if the workspace doesn't exist, regardless of what Sam returns from the permission check. Consider the use case that triggered this PR: deleting a non-existent workspace. In that use case, Sam returns a 403 since it can find no permission for the user. However, we want to return 404. I implemented this logic simply by  putting `getWorkspace()` first, letting it throw 404 if the workspace doesn't exist,  then performing the authz check which  will throw 403  or pass.

File: src/main/java/bio/terra/workspace/db/WorkspaceDao.java (hunk 1 above)
Comment (davidangb) [reply src/main/java/bio/terra/workspace/db/WorkspaceDao.java]: instead of making this an enum, I have clarified its usage in the javadoc, and renamed the argument to reflect its true usage. Does that work?

File: src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java (hunk 2 above)
Comment (davidangb) [reply src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java]: renamed to `validateWorkspaceAndAction`, how's that? The codebase uses "action" consistently instead of "operation" for the thing in question here.

**Review by ungwudik — APPROVED** [review]
Looks good to me.

**Review by zloery — APPROVED** [review]
new changes lgtm

File: src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java (hunk 2 above)
Comment (jaycarlton) [reply src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java]: sounds good

File: src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java (hunk 4 above)
Comment (jaycarlton) [reply src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java:92]: Please add a comment to that effect.

File: src/main/java/bio/terra/workspace/db/WorkspaceDao.java (hunk 1 above)
Comment (jaycarlton) [reply src/main/java/bio/terra/workspace/db/WorkspaceDao.java]: sure. we could change later if we find we want a pie chart of actions or something.
